catanatron_core/catanatron/state.py

```diff
@@ -76,6 +76,10 @@
     # de-normalized features (for performance since we think they are good features)
     "ACTUAL_VICTORY_POINTS": 0,
     "LONGEST_ROAD_LENGTH": 0,
+    "KNIGHT_PURCHASED_THIS_TURN": 0,
+    "MONOPOLY_PURCHASED_THIS_TURN": 0,
+    "YEAR_OF_PLENTY_PURCHASED_THIS_TURN": 0,
+    "ROAD_BUILDING_PURCHASED_THIS_TURN": 0,
```

Comment:
Can you share more about the motivation of these? I believe this is what the `_HAS_PLAYED_DEVELOPMENT_CARD_IN_TURN` state flag is going after (to know if player can play again). Could we use the existing `_HAS_PLAYED_DEVELOPMENT_CARD_IN_TURN` instead?

Reply:
If a player has a knight card at the start of their turn, then purchases 2 knight cards, they should still be able to play the one knight card. 
If a player has no knight card at the start of their turn, then purchases 2 knight cards, they should not be able to play them in the same turn.
This is to disallow them from purchasing/using on the same turn. The other way to handle it would be to track which dev cards they started with, but this seemed more intuitive to me at the time. 

Alternatively, we could have something like KNIGHT_OWNED_AT_START_OF_TURN, and check in player_can_play_dev() for:
{ KNIGHT_OWNED_AT_START_OF_TURN && !HAS_PLAYED_DEVELOPMENT_CARD_IN_TURN }

Looking back, maybe is more straightforward